fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,375 @@
+    pub fn add_vote(
+        &self,
+        validator_key: K,
+        block_height: BlockHeight,
+        block_hash: V,
+    ) -> StmResult<bool, Error<K, V>> {
+        if *self.pause_votes.read()? {
+            retry()?;
+        }
+
+        let min_height = self.last_finalized_height()?;
+
+        if block_height < min_height {
+            return Ok(false);
+        }
+
+        if !self.has_power(&validator_key)? {
+            return abort(Error::UnpoweredValidator(validator_key));
+        }
+
+        let mut votes = self.votes.read_clone()?;
+        let votes_at_height = votes.entry(block_height).or_default();
+
+        for (bh, vs) in votes_at_height.iter() {
+            if *bh != block_hash && vs.contains(&validator_key) {
+                return abort(Error::Equivocation(
+                    validator_key,
+                    block_height,
+                    block_hash,
+                    bh.clone(),
+                ));
+            }
+        }
+
+        let votes_for_block = votes_at_height.entry(block_hash).or_default();
+
+        if votes_for_block.insert(validator_key).is_some() {
+            return Ok(false);
+        }
+
+        self.votes.write(votes)?;
+
+        Ok(true)
+    }
+
+    /// Pause adding more votes until we are finished calling `find_quorum` which
+    /// automatically re-enables them.
+    pub fn pause_votes_until_find_quorum(&self) -> Stm<()> {
+        self.pause_votes.write(true)
+    }
+
+    /// Find a block on the (from our perspective) finalized chain that gathered enough votes from validators.
+    pub fn find_quorum(&self) -> Stm<Option<(BlockHeight, V)>> {
+        self.pause_votes.write(false)?;
+
+        let quorum_threshold = self.quorum_threshold()?;
+        let chain = self.chain.read()?;
+
+        let Some((finalized_height, _)) = chain.get_min() else {
+            return Ok(None);
+        };
+
+        let votes = self.votes.read()?;
+        let power_table = self.power_table.read()?;
+
+        let mut weight = 0;
+        let mut voters = im::HashSet::new();
+
+        for (block_height, block_hash) in chain.iter().rev() {
+            if block_height == finalized_height {
+                break;
+            }
+            let Some(block_hash) = block_hash else {
+                continue; // Skip null blocks
+            };
+            let Some(votes_at_height) = votes.get(block_height) else {
+                continue;
+            };
+            let Some(votes_for_block) = votes_at_height.get(block_hash) else {
+                continue; // We could detect equovicating voters here.
+            };
+
+            for vk in votes_for_block {
+                if voters.insert(vk.clone()).is_none() {
+                    // New voter, get their current weight; it might be 0 if they have been removed.
+                    weight += power_table.get(vk).cloned().unwrap_or_default();
+                }
+            }
```

Comment:
Do validators gossip about every single block they've observed? What if a validator goes down or has their parent sync loop interrupted, and then syncs back to parent HEAD in a batch? e.g. going for epoch 100 to 200 immediately. Will it send 100 gossip messages? (which may be considered spamming) Or will sending a vote for the latest implicitly vote for the rest of the chain? I think we should do the latter, but it's not clear to me if that's what's implemented here.

Reply:
Great question; the mechanism that should prevent us from spamming everyone with votes on blocks in the past is the the [TendermintAwareSyncer](https://github.com/consensus-shipyard/ipc/blob/335748d945066123629f60285e2890e34bfbaa82/fendermint/vm/topdown/src/sync/tendermint.rs#L30-L31) which checks if CometBFT is syncing with the network. If it is, it skips syncing with the parent in that round altogether, expecting to see final decisions in the blocks instead. 

Because the parent syncer doesn't run, there are no blocks added by it to the tally either, so we won't vote on anything.

Reply:
If the parent were CometBFT with a 1s block time, we should probably avoid gossiping on each height!

> syncs back to parent HEAD in a batch? e.g. going for epoch 100 to 200 immediately. Will it send 100 gossip messages?

That depends on whether CometBFT itself lost its sync with the subnet chain: https://docs.cometbft.com/v0.37/core/block-sync

It has fast sync, but on the bottom of the page that going _back_ to fast sync is an open issue.

Reply:
https://linear.app/interplanetary-consensus/issue/ENG-624/throttle-vote-gossiping